src/utils/message_entity.rs

```diff
@@ -0,0 +1,116 @@
+use std::string::FromUtf16Error;
+
+use teloxide_core::types::MessageEntity;
+
+#[derive(Debug)]
+pub enum InvokeEntityError {
+    TooShortText { offset: usize, length: usize },
+    FromUTF16(FromUtf16Error),
+}
+
+impl From<FromUtf16Error> for InvokeEntityError {
+    fn from(e: FromUtf16Error) -> Self {
+        Self::FromUTF16(e)
+    }
+}
+
+impl std::fmt::Display for InvokeEntityError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            InvokeEntityError::TooShortText { offset, length } => write!(
+                f,
+                "Text is too short for entity with offset {} and length {}",
+                offset, length
+            ),
+            InvokeEntityError::FromUTF16(e) => e.fmt(f),
+        }
+    }
+}
+
+impl std::error::Error for InvokeEntityError {}
+
+fn invoke_entity_from_utf16(
+    text: &[u16],
+    entity: &MessageEntity,
+) -> Result<String, InvokeEntityError> {
+    let start = entity.offset;
+    let end = entity.offset + entity.length;
+    if text.len() < end {
+        Err(InvokeEntityError::TooShortText { offset: entity.offset, length: entity.length })
+    } else {
+        String::from_utf16(&text[start..end]).map_err(Into::into)
+    }
+}
+pub fn invoke_entity(text: &str, entity: &MessageEntity) -> Result<String, InvokeEntityError> {
+    let text: Vec<_> = text.encode_utf16().collect();
+    invoke_entity_from_utf16(text.as_slice(), entity)
+}
+pub fn invoke_entities(
+    text: &str,
+    entities: &Vec<MessageEntity>,
+) -> Result<Vec<String>, InvokeEntityError> {
+    let text16: Vec<_> = text.encode_utf16().collect();
+    let slice = text16.as_slice();
+    let mut result = Vec::new();
+    for e in entities {
+        result.push(invoke_entity_from_utf16(slice, e)?);
+    }
```

Comment:
This is extremely inefficient.
At the very least it converts the same string to utf-16 on every iteration.

I think we can implement what we want in a single iteration over `text`'s `char`s, see this example: [(playground)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0820c88deef889eb108f82ab92c885b7).

Reply:
no, it doesn't. In this case we use prepared utf16 slice as function argument. 

and there we use String::from_utf16 as many as count of entities. And final strings - just what we need in place of usage (not ranges, i think).

about playground example: it's too complexity to understand. IMHO simple and transparent code is better than ununderstandable and littlebit faster. Overrall it's looks like premature optimization.

Reply:
> In this case we use prepared utf16 slice as function argument.

My bad, I for some reason thought that this calls `invoke_entity` >_>

Reply:
> And final strings - just what we need in place of usage (not ranges, i think).

Only for some use-cases.